File: cordapp/src/main/java/net/corda/derivativestradingnetwork/TokenTransferFlow.java
package net.corda.derivativestradingnetwork;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.contracts.CommandData;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.flows.*;
import net.corda.core.identity.Party;
import net.corda.core.node.services.NetworkMapCache;
import net.corda.core.node.services.VaultService;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.core.utilities.ProgressTracker;
import net.corda.derivativestradingnetwork.states.MoneyToken;

import java.util.Currency;
import java.util.List;
import java.util.stream.Collectors;

@InitiatingFlow
@StartableByRPC
public class TokenTransferFlow extends FlowLogic<Void> {
    private final Currency currency;
    private final int amount;
    private final Party newOwner;

    private final ProgressTracker progressTracker = new ProgressTracker();

    public TokenTransferFlow(Currency currency, int amount, Party newOwner) {
        this.currency = currency;
        this.amount = amount;
        this.newOwner = newOwner;
    }

    @Override
    public ProgressTracker getProgressTracker() {
        return progressTracker;
    }

    @Suspendable
    @Override
    public Void call() throws FlowException {
        NetworkMapCache networkMap = getServiceHub().getNetworkMapCache();
        Party notary = networkMap.getNotaryIdentities().get(0);

        VaultService vault = getServiceHub().getVaultService();
        List<StateAndRef<MoneyToken.State>> stateAndRefs =
                vault.queryBy(MoneyToken.State.class).getStates();

        List<StateAndRef<MoneyToken.State>> matchingStates = stateAndRefs.stream().filter(stateAndRef -> {
            MoneyToken.State state = stateAndRef.getState().getData();
            Boolean isCorrectAmount = state.getAmount() == amount;
            Boolean isCorrectCurrency = state.getCurrency().equals(currency);
            return isCorrectAmount && isCorrectCurrency;
        }).collect(Collectors.toList());

        if (matchingStates.isEmpty())
            throw new IllegalArgumentException("No matching token to transfer.");
        StateAndRef<MoneyToken.State> inputStateAndRef = matchingStates.get(0);
        MoneyToken.State inputState = inputStateAndRef.getState().getData();

        MoneyToken.State outputToken =
                new MoneyToken.State(amount, currency, inputState.getIssuer(), newOwner, null);
        CommandData transferCommand = new MoneyToken.Commands.Transfer();

        TransactionBuilder txBuilder = new TransactionBuilder(notary);
        txBuilder.addInputState(inputStateAndRef);
        txBuilder.addOutputState(outputToken, MoneyToken.Companion.getCONTRACT_NAME());
        txBuilder.addCommand(transferCommand, getOurIdentity().getOwningKey());

        txBuilder.verify(getServiceHub());

        SignedTransaction stx = getServiceHub().signInitialTransaction(txBuilder);

        subFlow(new FinalityFlow(stx));

        return null;
    }
}

File: cordapp/src/main/kotlin/net/corda/derivativestradingnetwork/flow/TokenTransferFlow.kt
package net.corda.derivativestradingnetwork.flow

import co.paralleluniverse.fibers.Suspendable
import com.google.common.collect.ImmutableList
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.transactions.TransactionBuilder
import net.corda.core.utilities.ProgressTracker
import net.corda.derivativestradingnetwork.states.MoneyToken
import java.security.PublicKey
import java.util.*

@InitiatingFlow
@StartableByRPC
class TokenTransferFlow(private val currency: Currency, private val amount: Int, private val newOwner: Party) : FlowLogic<Unit>() {

    override val progressTracker: ProgressTracker? = ProgressTracker()

    @Suspendable
    override fun call(): Unit {
        val networkMap = serviceHub.networkMapCache
        val notary = networkMap.notaryIdentities[0]

        val vault = serviceHub.vaultService
        val stateAndRefs = vault.queryBy(MoneyToken.State::class.java).states

        val matchingStates = stateAndRefs.filter { (state) ->
            val state = state.data
            val isCorrectAmount = state.amount == amount.toLong()
            val isCorrectCurrency = state.currency == currency
            isCorrectAmount && isCorrectCurrency
        }

        if (matchingStates.isEmpty())
            throw IllegalArgumentException("No matching token to transfer.")
        val inputStateAndRef = matchingStates[0]
        val (_, _, issuer, _, amlAuthority) = inputStateAndRef.state.data

        val outputToken = MoneyToken.State(amount.toLong(), currency, issuer, newOwner, amlAuthority)
        val transferCommand = MoneyToken.Commands.Transfer()

        val requiredSigners: List<PublicKey>
        if (amount > 150) {
            requiredSigners = Arrays.asList(ourIdentity.owningKey, outputToken.amlAuthority.owningKey)
        } else {
            requiredSigners = listOf(ourIdentity.owningKey)
        }

        val txBuilder = TransactionBuilder(notary)
                .addInputState(inputStateAndRef)
                .addOutputState(outputToken, MoneyToken.CONTRACT_NAME)
                .addCommand(transferCommand, requiredSigners)

        txBuilder.verify(serviceHub)

        val partSignedTx = serviceHub.signInitialTransaction(txBuilder)

        if (amount > 150) {
            val sessionWithAml = initiateFlow(outputToken.amlAuthority)
            val fullySignedTx = subFlow(CollectSignaturesFlow(partSignedTx, ImmutableList.of(sessionWithAml)))
            subFlow(FinalityFlow(fullySignedTx))
        } else {
            subFlow(FinalityFlow(partSignedTx))
        }
    }
}